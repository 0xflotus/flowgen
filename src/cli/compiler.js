/* @flow */
import { createSourceFile, ScriptTarget } from "typescript";
import fs from "fs";
import tsc from "typescript-compiler";

import namespaceManager from "../namespaceManager";
import { type Options, assignOptions, resetOptions } from "../options";
import { recursiveWalkTree } from "../parse";

const compile = (sourceFile): string => {
  const rootNode = recursiveWalkTree(sourceFile);

  const output = rootNode
    .getChildren()
    .map(child => {
      return child.print();
    })
    .join("");

  return output;
};

/**
 * Compiles typescript files
 */
export default {
  reset(options?: Options) {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();
  },

  compile,

  compileTest: (path: string, target: string): void => {
    tsc.compile(path, "--module commonjs -t ES6 --out " + target);
  },

  compileDefinitionString: (string: string, options?: Options): string => {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();

    return compile(
      createSourceFile("/dev/null", string, ScriptTarget.Latest, false),
    );
  },

  compileDefinitionFile: (path: string, options?: Options): string => {
    resetOptions();
    if (options) {
      assignOptions(options);
    }
    namespaceManager.reset();

    return compile(
      createSourceFile(
        path,
        fs.readFileSync(path).toString(),
        ScriptTarget.Latest,
        false,
      ),
    );
  },
};
