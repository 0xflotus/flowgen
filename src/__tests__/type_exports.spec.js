// @flow
import compiler from "../cli/compiler";
import beautify from "../cli/beautifier";

it("should handle exported types", () => {
  const ts = "export declare type FactoryOrValue<T> = T | (() => T);";
  expect(beautify(compiler.compileDefinitionString(ts))).toMatchSnapshot();
});
