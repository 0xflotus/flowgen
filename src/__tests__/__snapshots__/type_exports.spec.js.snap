// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle exported types 1`] = `
"export type FactoryOrValue<T> = T | (() => T);
"
`;
